### ISSUE TITLE
Incorrectly computed dependency versions (under Pipfile.lock) for --dev

### ISSUE THREAD

**Issue by mirceaulinic** [pypa/pipenv#5107]
### Issue description

I'm running an environment where, under `dev`, one package has a specific constraint around the version of another package. More specifically, consider the following Pipfile:

```pipfile
[[source]]
url = "https://pypi.org/simple"
verify_ssl = true
name = "pypi"

[packages]
salt = "==3002.8"
napalm = "==3.3.1"
junos-eznc = "==2.6.3"
capirca = "==2.0.4"

[dev-packages]
docker-compose = ">=1.29.2"

[requires]
python_version = "3.9"
```

Some of the packages under `[packages]` require PyYAML>=5.1. So any recent version would work. But under dev, `docker-compose` needs PyYAML<6.0.

### Expected result

When running `pipenv lock --dev` I would expect that (at least under `dev`), Pipenv would compute that PyYAML should be pinned to 5.4.1 (which would respect both >=5.1 and <6.0 conditions).

### Actual result

Under Pipfile.lock, it pins PyYAML to 6.0 for both `default` and `dev`. Please check below, the output from `pipenv --support`.

Because of this, when dumping the requirements to a file and installing it causes the following failure:

```
The conflict is caused by:
    The user requested pyyaml==6.0
    capirca 2.0.4 depends on PyYAML
    junos-eznc 2.6.3 depends on PyYAML>=5.1
    napalm 3.3.1 depends on pyYAML
    salt 3002.8 depends on PyYAML
    yamlordereddictloader 0.4.0 depends on pyyaml
    docker-compose 1.29.2 depends on PyYAML<6 and >=3.10
```

When installing via `pipenv install --dev` it doesn't fail, but doesn't honour the <6 constraint, as it goes ahead and incorrectly installs `PyYAML==6.0`.

Adding `PyYAML = "==5.4.1"` under `[dev-packages]` doesn't have any effect, as it seems that the version computed for `default` is simply copied to `dev` without compatibility checks. 

### Steps to replicate

Provide the steps to replicate (which usually at least includes the commands and the Pipfile).

```bash
$ pipenv lock --dev
$ pipenv requirements --dev > /tmp/req.txt
$ pip install -r /tmp/req.txt
```

-------------------------------------------------------------------------------

The output of `pipenv --support` is too long (GitHub returned `There was an error creating your Issue: body is too long (maximum is 65536 characters).`) but can be found [here](https://gist.github.com/mirceaulinic/6746ea681cf97ee58ca4877b4e80c265).

**Comment by matteius** [comment]
@mirceaulinic and @Mulanir 
  At first this sounded like a repeat report of https://github.com/pypa/pipenv/issues/4665 however I see you are using the new requirements command.  When you pass `--dev` to the new `requirements` command, it tries to include both the dev and regular packages which in your case has conflicts -- and that may fall under the purview of the above bug report.  If you wanted to generate the requirements of just the dev requirements -- you would pass `--dev-only` so `pipenv requirements --dev-only`

Now if you are thinking I want both the dev and regular packages in your requirements txt, then I think there are going to be problems because of the conflict in resolution.   It appears the requirements command when using the `--dev` flag will update the requirements and override any specifiers from the main packages with the dev-packages, see: https://github.com/pypa/pipenv/blob/main/pipenv/cli/command.py#L779-L782  cc: @ImreC 

So I think that this issue as it pertains to the requirements command is a similar issue to the one I linked above.  The problem with that issue is that there was never any intent defined from the original pipenv authors on what the expected behavior should be when there is the same package in packages and dev-packages.  I don't believe that trying to resolve a version that works for both sections is the right answer either, because in some cases there will never be matching version that can satisfy both sets of constraints.  I would advice trying to not end up in the situation where your development dependencies require a different version of a package than is compatible with your production dependencies.  Otherwise I think you need to consider having two different virtualenvs -- one you install the production dependencies to and one that you install the dev dependencies too.

I do wonder if in the code I linked to for the `requirements` command, if we should have this logic first add the dev dependencies so that the production dependencies can overwrite any of those? -- that seems more correct to me than allowing the dev dependencies to break the production specifiers.

**Comment by matteius** [comment]
This was fixed in `pipenv=2022.7.4` as it changed the order (regular packages overwrite dev packages and not the other way around).

**Comment by mirceaulinic** [comment]
That's great, thank you so much for the update @matteius! (sorry I missed your previous message)
